### PR TITLE
Add logs for why start was blocked

### PR DESCRIPTION
**Pull request by LoganxDev**
#2272

### CODE REVIEW

**Comment by pd0wm** [comment]
That's what I had in mind. Can we clean up the code why introducing a helper function? Something like

```python
def block_startup(should_start, condition, reason):
  if should_start and not condition:
    should_start = False
    cloudlog.info(f"Blocking startup: {reason}")
  
  return should_start

...

should_start = block_startup(should_start, msg.thermal.freeSpace <= 0.02, "<2 percent memory")
```

File: selfdrive/thermald/thermald.py [hunk 1]
```diff
@@ -360,33 +360,46 @@ def thermald_thread():
...
     # confirm we have completed training and aren't uninstalling
-    should_start = should_start and accepted_terms and (not do_uninstall) and \
-                   (completed_training or current_branch in ['dashcam', 'dashcam-staging'])
-
-    # check for firmware mismatch
-    should_start = should_start and fw_version_match
+    if should_start and not (accepted_terms and (not do_uninstall) and
+      (completed_training or current_branch in ['dashcam', 'dashcam-staging'])):
+      cloudlog.info("Blocking start: Did not complete training or uninstalling")
+      should_start = False
```
Comment (pd0wm) [inline selfdrive/thermald/thermald.py:371]: Let's give uninstall it's own check, has nothing to do with the terms.

File: selfdrive/thermald/thermald.py [hunk 2]
```diff
@@ -156,6 +158,24 @@ def handle_fan_uno(max_cpu_temp, bat_temp, fan_speed, ignition):
   return new_speed
 
 
+def set_offroad_alert_if_changed(offroad_alert, show_alert):
+  if offroad_alert in prev_offroad_states and \
```
Comment (pd0wm) [inline selfdrive/thermald/thermald.py:162]:
```suggestion
  if prev_offroad_states.get(offroad_alert, None) == show_alert:
```

Comment (pd0wm) [reply selfdrive/thermald/thermald.py:162]: I guess your version is more explicit/pythonic. You can choose what you like better.

File: selfdrive/thermald/thermald.py [hunk 3]
```diff
@@ -360,51 +372,29 @@ def thermald_thread():
...
-    if should_start and not time_valid:
-      cloudlog.info("Blocking start: Invalid system time")
-      should_start = False
-
-    # don't start while taking snapshot
+    should_start = startup_allowed(should_start, (msg.thermal.freeSpace > 0.02), "<2 percent memory")
+
+    should_start = startup_allowed(should_start, (not do_uninstall), "Uninstalling")
+
+    should_start = startup_allowed(should_start,
```
Comment (pd0wm) [inline selfdrive/thermald/thermald.py:379]: Split up in accepted terms and completed training. You can also remove whitespace between these checks.

Comment (LoganxDev) [reply selfdrive/thermald/thermald.py:379]: Sounds good, should the dashcam branch check have its own error log?

File: selfdrive/thermald/thermald.py [hunk 4]
```diff
@@ -156,6 +158,23 @@ def handle_fan_uno(max_cpu_temp, bat_temp, fan_speed, ignition):
...
 
 
+def set_offroad_alert_if_changed(offroad_alert, show_alert, extra_text=None):
+  if prev_offroad_states.get(offroad_alert, None) == show_alert:
+    return
+  prev_offroad_states[offroad_alert] = show_alert
+  set_offroad_alert(offroad_alert, show_alert, extra_text)
+
+
+def startup_allowed(should_start, condition, reason, offroad_alert=None):
```
Comment (adeebshihadeh) [inline selfdrive/thermald/thermald.py:168]: this seems more complicated than it needs to be. can't you just leave the alerts how they are and build a dict wtih all the things we check? then should start is as simple as `should_start = all(startup_conditions.values())` and if it's blocked log the whole dict.

Comment (LoganxDev) [reply selfdrive/thermald/thermald.py:168]: I like the condition dict as well, I do think the offroad_alert wrapper should stay to clean up all the prev value checks though.

File: selfdrive/thermald/thermald.py (hunk 3 above)
Comment (pd0wm) [reply selfdrive/thermald/thermald.py:378]: It's 

```python
should_start = startup_allowed(should_start, accepted_terms, "Terms not accepted")
should_start = startup_allowed(should_start, (completed_training or current_branch in ['dashcam', 'dashcam-staging']), "Training not completed")
```

**Comment by pd0wm** [comment]
you can run `pre-commit` locally

File: selfdrive/thermald/thermald.py [hunk 5]
```diff
@@ -156,6 +158,14 @@ def handle_fan_uno(max_cpu_temp, bat_temp, fan_speed, ignition):
   return new_speed
 
 
+def set_offroad_alert_if_changed(offroad_alert, show_alert, extra_text=None):
+  global prev_offroad_states
```
Comment (pd0wm) [inline selfdrive/thermald/thermald.py:162]: I don't think you need `global`

Comment (LoganxDev) [reply selfdrive/thermald/thermald.py]: Ah, yes you're right.

File: selfdrive/thermald/thermald.py [hunk 6]
```diff
@@ -403,6 +393,7 @@ def thermald_thread():
         started_seen = True
         os.system('echo performance > /sys/class/devfreq/soc:qcom,cpubw/governor')
     else:
+      cloudlog.info(f"Start was blocked, startup conditions: {str(startup_conditions)}")
```
Comment (adeebshihadeh) [inline selfdrive/thermald/thermald.py:396]: don't think str is needed with f strings

Comment (adeebshihadeh) [inline selfdrive/thermald/thermald.py:396]:
```suggestion
      cloudlog.event("Startup blocked", startup_conditions=startup_conditions)
```

File: selfdrive/thermald/thermald.py [hunk 7]
```diff
@@ -328,70 +342,46 @@ def thermald_thread():
...
-      should_start = should_start and (not is_taking_snapshot) and (not is_viewing_driver)
-
-    if fw_version_match and not fw_version_match_prev:
-      set_offroad_alert("Offroad_PandaFirmwareMismatch", False)
-    if not fw_version_match and fw_version_match_prev:
-      set_offroad_alert("Offroad_PandaFirmwareMismatch", True)
-
+    startup_conditions["free_space"] = msg.thermal.freeSpace > 0.02
+    startup_conditions["completed_training"] = completed_training or (current_branch in ['dashcam', 'dashcam-staging'])
+    startup_conditions["not_driver_view"] = params.get("IsDriverViewEnabled") == b"1"
```
Comment (pd0wm) [inline selfdrive/thermald/thermald.py:378]:
```suggestion
    startup_conditions["not_driver_view"] = not params.get("IsDriverViewEnabled") == b"1"
```

File: selfdrive/thermald/thermald.py [hunk 8]
```diff
@@ -328,70 +342,46 @@ def thermald_thread():
...
+    startup_conditions["not_taking_snapshot"] = params.get("IsTakingSnapshot") == b"1"
```
Comment (pd0wm) [inline selfdrive/thermald/thermald.py:379]:
```suggestion
    startup_conditions["not_taking_snapshot"] = not params.get("IsTakingSnapshot") == b"1"
```

File: selfdrive/thermald/thermald.py [hunk 9]
```diff
@@ -403,6 +393,7 @@ def thermald_thread():
         started_seen = True
         os.system('echo performance > /sys/class/devfreq/soc:qcom,cpubw/governor')
     else:
+      cloudlog.event("Startup blocked", startup_conditions=startup_conditions)
```
Comment (pd0wm) [inline selfdrive/thermald/thermald.py:396]:
```suggestion
    if startup_conditions["ignition"]:
      cloudlog.event("Startup blocked", startup_conditions=startup_conditions)
```

File: selfdrive/thermald/thermald.py [hunk 10]
```diff
@@ -403,6 +393,8 @@ def thermald_thread():
         started_seen = True
         os.system('echo performance > /sys/class/devfreq/soc:qcom,cpubw/governor')
     else:
+    if startup_conditions["ignition"]:
+      cloudlog.event("Startup blocked", startup_conditions=startup_conditions)
```
Comment (pd0wm) [inline selfdrive/thermald/thermald.py:397]:
```suggestion
      if startup_conditions["ignition"]:
        cloudlog.event("Startup blocked", startup_conditions=startup_conditions)
```

File: selfdrive/thermald/thermald.py (hunk 6 above)
Comment (pd0wm) [inline selfdrive/thermald/thermald.py]:
```suggestion
      cloudlog.info(f"Start was blocked, startup conditions: {startup_conditions}")
```

**Comment by pd0wm** [comment]
Doing some final cleanup of the update alerts here: https://github.com/commaai/openpilot/pull/2346

**Pull request closed without merging**